crates/sui-core/src/traffic_controller.rs

```diff
@@ -0,0 +1,178 @@
+            (Some(connection_ip), _) => self.check_and_clear_blocklist(connection_ip, true).await,
+            (_, Some(proxy_ip)) => self.check_and_clear_blocklist(proxy_ip, false).await,
+            _ => true,
+        }
+    }
+
+    async fn check_and_clear_blocklist(&self, ip: SocketAddr, connection_ips: bool) -> bool {
+        let blocklist = if connection_ips {
+            self.blocklists.connection_ips.clone()
+        } else {
+            self.blocklists.proxy_ips.clone()
+        };
+
+        let now = Utc::now();
+        let expiration = blocklist.read().get(&ip).copied();
+        match expiration {
+            Some(expiration) if now >= expiration => {
+                blocklist.write().remove(&ip);
+                true
+            }
+            None => true,
+            _ => false,
+        }
+    }
+}
+
+async fn run_tally_loop(
+    mut receiver: mpsc::Receiver<TrafficTally>,
+    policy_config: PolicyConfig,
+    blocklists: Blocklists,
+) {
+    let mut spam_policy = policy_config.clone().to_spam_policy();
+    let mut error_policy = policy_config.clone().to_error_policy();
+    let spam_blocklists = Arc::new(blocklists.clone());
+    let error_blocklists = Arc::new(blocklists);
+    loop {
+        tokio::select! {
+            received = receiver.recv() => match received {
+                Some(tally) => {
+                    handle_spam_tally(&mut spam_policy, &policy_config, tally.clone(), spam_blocklists.clone()).await;
+                    handle_error_tally(&mut error_policy, &policy_config, tally, error_blocklists.clone()).await;
+                }
+                None => {
+                    panic!("TrafficController tally channel closed unexpectedly");
+                },
+            }
+        }
+    }
+}
+
+async fn handle_error_tally(
+    policy: &mut TrafficControlPolicy,
+    config: &PolicyConfig,
+    tally: TrafficTally,
+    blocklists: Arc<Blocklists>,
+) {
+    let err = if let Some(err) = tally.clone().result.err() {
+        err
+    } else {
+        return;
+    };
+    if config.tallyable_error_codes.contains(&err) {
+        handle_tally_impl(policy, config, tally, blocklists).await
+    }
+}
+
+async fn handle_spam_tally(
+    policy: &mut TrafficControlPolicy,
+    config: &PolicyConfig,
+    tally: TrafficTally,
+    blocklists: Arc<Blocklists>,
+) {
+    // TODO -- add update of spam blocklist
+    handle_tally_impl(policy, config, tally, blocklists).await
+}
+
+async fn handle_tally_impl(
+    policy: &mut TrafficControlPolicy,
+    config: &PolicyConfig,
+    tally: TrafficTally,
+    blocklists: Arc<Blocklists>,
+) {
+    let PolicyResponse {
+        block_connection_ip,
+        block_proxy_ip,
+    } = policy.handle_tally(tally.clone());
+    if block_connection_ip {
+        blocklists.connection_ips.write().insert(
+            tally
+                .connection_ip
+                .expect("Expected connection IP if policy is blocking it"),
```

Comment:
it would be better to make PolicyResponse an enum so that each case can have exactly the state that is required, rather than using `expect`

Reply:
Thanks, not sure why I didn't do that